fabric-network/src/impl/query/roundrobinqueryhandler.js

```diff
@@ -13,52 +13,46 @@ const util = require('util');
 const logger = require('fabric-network/lib/logger').getLogger('RoundRobinQueryHandler');
 
 class RoundRobinQueryHandler {
-	constructor(peers, options = {}) {
+	constructor(peers) {
 		logger.debug('constructor: peers=%j', peers.map((peer) => peer.name));
 		this._peers = peers;
 		this._currentPeerIndex = 0;
-		this._options = options;
 	}
 
 	async evaluate(query) {
 		const method = 'evaluate';
 		logger.debug('%s - start', method);
 
-		const startPeerIndex = this._currentPeerIndex;
-		const errorMessages = [];
-		const options = {requestTimeout: 3000};
-		// use the timeout as the requestTimeout or let default
-		if (Number.isInteger(this._options.timeout)) {
-			options.requestTimeout = this._options.timeout * 1000; // in ms;
+		if (this._currentPeerIndex >= this._peers.length) {
+			// start over
+			this._currentPeerIndex = 0;
 		}
 
+		const startPeerIndex = this._currentPeerIndex++;
+		const errorMessages = [];
+
 		for (let i = 0; i < this._peers.length; i++) {
 			const peerIndex = (startPeerIndex + i) % this._peers.length;
-			this._currentPeerIndex = peerIndex;
-			const peer = this._peers[peerIndex];
-
-			logger.debug('%s - query sending to peer %s', method, peer.name);
-			const results = await query.send({targets:[peer]}, options);
 
-			if (results.errors.length > 0) {
-				logger.error('%s - problem with query to peer %s error:%s', method, peer.name, results.errors[0]);
-				// since only one peer, only one error
-				errorMessages.push(results.errors[0].message);
-				continue;
-			}
-
-			const endorsementResponse = results.responses[0];
+			const peer = this._peers[peerIndex];
+			logger.debug('%s - sending to peer %s', method, peer.name);
 
-			if (!endorsementResponse.endorsement) {
-				logger.debug('%s - peer response status: %s message: %s',
-					method,
-					endorsementResponse.response.status,
-					endorsementResponse.response.message);
-				throw new Error(endorsementResponse.response.message);
+			const results = await query.evaluate([peer]);
+			const result = results[peer.name];
+			if (result instanceof Error) {
+				errorMessages.push(result.toString());
+			} else {
+				if (result.status < 400) {
+					logger.debug('%s - return peer response status: %s', method, result.status);
+					return result.payload;
+				} else if (result.status < 500) {
+					logger.debug('%s - throw peer response status: %s message: %s', method, result.status, result.message);
+					throw Error(result.message);
+				} else {
+					logger.debug('%s - peer response message: %s', method, result.message);
+					errorMessages.push(result.message);
```

Comment:
This still isn't behaving as it used to. Any proposal response that wasn't endorsed (typically status >= 400) was treated as a "valid" error and thrown. No proposal responses were added to errorMessages and caused the code to go around the for-loop again.

Reply:
Right, I was not trying make it behave as before, but to consider the information provided by the high level query.  Where:
if we get less than 400 it will be considered good and stop looking and return the payload, this is will include the usual case of 200 and all the 300's which most consider to be good
if we get less than 500 it will be considered bad and stop looking and return error with message, this will include the 404 case of not found, which most consider to be bad and all peers should have the same results due to event handling.
if we get 500 or greater it will be considered bad and keep looking and only return the message if nothing else found, this will include cases where the container is down or the chaincode is not running and we should be trying other peers.

Reply:
If we want to change the default handler behaviour, we should do this consistently across the Node and Java SDK, not as part of a refactor of the low-level implementation